CI: use strict mode for compiling

zerbina opened the pull request:
## Summary

Extend `nimStrictMode` to error on unused imports, and enable
`nimStrictMode` when compiling the compiler and core tools during CI
runs.

The idea is to have this help with catching bugs and preventing
leftover debug code being merged, while not complicating local
development. Over time, `nimStrictMode` should promote more warnings
and hints to errors.

## Details

* treat `UnusedImport` warnings as an error when `nimStrictMode` is
  enabled
* remove the obsolete `runCI` koch command
* introduce the new `all-strict` koch command, which does the same
  as `all`, except that it enables strict mode
* when building the compiler and tools in strict mode, use an
  `--errorMax` of 3, so that compilation doesn't abort after the first
  error
* run `all-strict` instead of `all` in the CI

Due to a bug, the current csources compiler doesn't properly support
warnings/hints promoted to errors, so `nimStrictMode` is turned into a
no-op when `nimKochBootstrap` is defined (this define is set for the
first iteration of bootstrapping). This is a temporary workaround,
which can be removed after the next csources update.

zerbina committed:
config: treat `UnusedImport` as an error in strict mode

zerbina committed:
koch: repurpose the outdated `runci` command

It performs the same action as `all` now, but enables stricter
compilation options (`-d:nimStrictMode`).

zerbina committed:
ci: use `runCI` instead of `all` for booting

zerbina committed:
config: disable `nimStrictMode` for the first iteration

The csources compiler doesn't properly support warnings/hints promoted
to errors.

saem commented on tools/koch/koch.nim:
```diff
@@ -529,32 +531,13 @@ proc testTools(cmd: string) =
-  when not defined(bsd):
-    # the BSDs are overwhelmed already, so only run this test on the other machines:
-    kochExecFold("Boot Nim ORC", "boot -d:release --gc:orc --lib:lib")
+proc runCI() =
```
might be a better name given the repurposing -- soft suggestion.

```suggestion
proc ciBuild() =
```

saem reviewed (approved):
I like the change overall, especially as it drops a bunch of confusing deadcode along the way.

alaviss commented on tools/koch/koch.nim:
```diff
@@ -66,7 +66,9 @@ Boot options:
 
 Commands for core developers:
-  runCI                    runs continuous integration (CI), eg: github actions
+  runCI                    bootstrap the compiler and build tools for release,
+                           in the context of continuous integration (CI),
+                           e.g.: github actions
```
I think we should call this `all-strict` in this case.

alaviss commented on tools/koch/koch.nim:
```diff
@@ -529,32 +531,13 @@ proc testTools(cmd: string) =
+  ## (=CI).
+  # enable strict mode in CI, which turns some warnings and hints into
+  # errors. In addition, don't abort after the first error
+  kochExec("boot -d:danger -d:nimStrictMode --errorMax:3")
+  buildTools("-d:nimStrictMode --errorMax:3")
```
Instead of this, I think having buildReleaseBinaries taking flags is a better idea. That'd allow the build target logic to stay in one place.

zerbina committed:
koch: rename `runCI` command to `all-strict`

In addition, replace the `runCI` procedure with re-using
`buildReleaseBinaries`.

zerbina commented:
I've renamed the command to `all-strict` and made it use `buildReleaseBinaries`.

alaviss reviewed (approved):
LGTM

zerbina commented:
/merge

saem commented:
/merge

saem commented:
whoops, didn't refresh the page with comments properly, now I'm curious how this might work.

saem commented:
/merge

saem commented:
/merge

alaviss commented:
You don't really have to do that, the script is tolerant to multiple requests.

But maybe I'll have to improve the feedback when you do it twice somehow

saem commented:
it seemed to have gotten stuck eventually, and when I checked branches there was nothing in the merge queue.

alaviss commented:
I think you found a Github issue. This PR should be in merge queue now instead of being stuck in auto-merge.
